web/web.go

```diff
@@ -84,34 +84,71 @@ func openWebBrowser(url string) error {
 	return exec.Command(cmd, args...).Start()
 }
 
+func startGrpcWebProxy(ctx *progctx.ProgCtx) error {
+	grpcWebProxyProc = exec.CommandContext(ctx, "grpcwebproxy", []string{
+		fmt.Sprintf("--backend_addr=localhost:%d", grpcWebProxyParams.grpcServicePort),
+		"--run_tls_server=false",
+		"--allow_all_origins",
+		"--server_http_max_read_timeout=1h",
+		"--server_http_max_write_timeout=1h",
+		fmt.Sprintf("--server_bind_address=%s", grpcWebProxyParams.serverBindAddress),
+		fmt.Sprintf("--server_http_debug_port=%d", grpcWebProxyParams.serverHttpDebugPort),
+	}...)
+
+	return grpcWebProxyProc.Start()
+}
+
 func assureGrpcWebProxyRunning(ctx *progctx.ProgCtx) error {
 	if grpcWebProxyProc == nil {
-		proc := exec.CommandContext(ctx, "grpcwebproxy", []string{
-			fmt.Sprintf("--backend_addr=localhost:%d", grpcWebProxyParams.grpcServicePort),
-			"--run_tls_server=false",
-			"--allow_all_origins",
-			"--server_http_max_read_timeout=1h",
-			"--server_http_max_write_timeout=1h",
-			fmt.Sprintf("--server_bind_address=%s", grpcWebProxyParams.serverBindAddress),
-			fmt.Sprintf("--server_http_debug_port=%d", grpcWebProxyParams.serverHttpDebugPort),
-		}...)
-
-		if err := proc.Start(); err != nil {
+		if err := startGrpcWebProxy(ctx); err != nil {
 			return err
 		}
 
-		grpcWebProxyProc = proc
+		simplelogger.Infof("grpcwebproxy started.")
+
 		ctx.WaitAdd("grpcwebproxy", 1)
 		go func() {
 			defer ctx.WaitDone("grpcwebproxy")
 
 			err := grpcWebProxyProc.Wait()
 			if err != nil && ctx.Err() == nil {
-				simplelogger.Errorf("grpcwebproxy exit unexpectedly: %v", err)
+				simplelogger.Warnf("grpcwebproxy exit unexpectedly: %v, try restarting ...", err)
+
+				tryKillExistingGrpcWebProxyProcess()
+
+				if err = startGrpcWebProxy(ctx); err != nil {
+					simplelogger.Errorf("grpcwebproxy restart failed: %v", err)
+					return
+				}
+
+				simplelogger.Infof("grpcwebproxy restarted.")
+				err = grpcWebProxyProc.Wait()
+
+				if err != nil && ctx.Err() == nil {
+					simplelogger.Errorf("grpcwebproxy exit unexpectedly: %v", err)
+					simplelogger.Errorf("Web visualization might not be working properly!")
+				}
 			}
 		}()
-		simplelogger.Infof("grpcwebproxy started.")
 	}
 
 	return nil
 }
+
+func tryKillExistingGrpcWebProxyProcess() {
+	var err error
+
+	defer func() {
+		if err != nil {
+			simplelogger.Warnf("Kill existing grpcwebproxy process failed: %v", err)
+		}
+	}()
+
+	killcmd := fmt.Sprintf("ps aux | grep -e \"grpcwebproxy.*\\-\\-server_http_debug_port=%d\" | awk '{print $2}' | xargs kill", grpcWebProxyParams.serverHttpDebugPort)
```

Comment:
can we stop `grpcwebproxy` with `pkill --signal SIGTERM grpcwebproxy`?

Reply:
what's the difference of the two commands?

Reply:
Thanks, I didn't know `pkill` command until now.

Reply:
Changed to use `pkill`. Argument `--signal SIGTERM` is omitted since it's the default.